ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifierImpl.java

```diff
@@ -220,14 +221,58 @@ private void internalForkChoiceUpdated(
 
     LOG.debug("internalForkChoiceUpdated forkChoiceUpdateData {}", forkChoiceUpdateData);
 
-    final Optional<UInt64> attributesSlot =
-        proposingSlot.or(() -> recentChainData.getCurrentSlot().map(UInt64::increment));
-
-    attributesSlot.ifPresent(this::updatePayloadAttributes);
+    calculatePayloadAttributesSlot(forkChoiceState, proposingSlot)
+        .ifPresent(this::updatePayloadAttributes);
 
     sendForkChoiceUpdated();
   }
 
+  /**
+   * Determine for which slot we should calculate payload attributes (block proposal)
+   *
+   * <pre>
+   * this will guarantee that whenever we calculate a payload attributes for a slot, it will remain stable until:
+   * 1. next slot attestation due is reached (internalAttestationsDue forcing attributes calculation for next slot)
+   * OR
+   * 2. we imported the block for current slot and has become the head
+   * </pre>
+   */
+  private Optional<UInt64> calculatePayloadAttributesSlot(
+      final ForkChoiceState forkChoiceState, final Optional<UInt64> proposingSlot) {
+    if (proposingSlot.isPresent()) {
+      // We are in the context of a block production, so we should use the proposing slot
+      return proposingSlot;
+    }
+
+    final Optional<UInt64> currentSlot = recentChainData.getCurrentSlot();
+    if (currentSlot.isEmpty()) {
+      // We are pre-genesis, so we don't care about proposing slots
+      return Optional.empty();
+    }
+
+    final Optional<UInt64> maybeCurrentPayloadAttributesSlot =
+        forkChoiceUpdateData
+            .getPayloadBuildingAttributes()
+            .map(PayloadBuildingAttributes::getProposalSlot);
+
+    if (maybeCurrentPayloadAttributesSlot.isPresent()
+        // we are still in the same slot as the last proposing slot
+        && currentSlot.get().equals(maybeCurrentPayloadAttributesSlot.get())
+        // we have not yet imported our own produced block
+        && forkChoiceState.getHeadBlockSlot().isLessThan(maybeCurrentPayloadAttributesSlot.get())) {
+
+      LOG.info("Returned same slot as current payload attributes", new Exception("Stack trace"));
```

Comment:
this for our debug?

Reply:
i cant really approve a log.info with `new Exception` :)

Reply:
Sure:) wasn't intended to be approved in this current form